### PR TITLE
GRIDEDIT-612 First implementation of snapping grid to land boundary

### PR DESCRIPTION
**Pull request by BillSenior**

### CODE REVIEW

File: libs/MeshKernel/src/CurvilinearGrid/CurvilinearGridSnapping.cpp [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+        constexpr UInt nump = 80;
+        auto m1 = static_cast<UInt>(std::max<int>(1, snappedNodeIndex.m_m + 1 - nump * m_smoothingRegionIndicator.m_m) - 1);
+        auto m2 = static_cast<UInt>(std::min<int>(m_grid.m_numM, snappedNodeIndex.m_m + 1 + nump * m_smoothingRegionIndicator.m_m) - 1);
+        auto n1 = static_cast<UInt>(std::max<int>(1, snappedNodeIndex.m_n + 1 - nump * m_smoothingRegionIndicator.m_n) - 1);
+        auto n2 = static_cast<UInt>(std::min<int>(m_grid.m_numN, snappedNodeIndex.m_n + 1 + nump * m_smoothingRegionIndicator.m_n) - 1);
+        return {CurvilinearGridNodeIndices(m1, n1), CurvilinearGridNodeIndices(m2, n2)};
+    }
+    else
+    {
+        auto m1 = static_cast<UInt>(std::max<int>(m_indexBoxLowerLeft.m_m, snappedNodeIndex.m_m + 1 - 10000 * m_smoothingRegionIndicator.m_m - 1));
```
Comment (lucacarniato) [inline libs/MeshKernel/src/CurvilinearGrid/CurvilinearGridSnapping.cpp]: Can we move 10000 as a static constant in the header file of CurvilinearGridSnapping?
Same applies to other numbers, such as nump, aspectRatio

**Review by lucacarniato — CHANGES_REQUESTED** [review]

File: libs/MeshKernel/include/MeshKernel/BoundingBox.hpp [hunk 2]
```diff
@@ -1,134 +1,173 @@
...
+
+inline meshkernel::BoundingBox meshkernel::Merge(const BoundingBox& b1, const BoundingBox& b2)
+{
+    Point lowerLeft{std::min(b1.lowerLeft().x, b2.lowerLeft().x), std::min(b1.lowerLeft().y, b2.lowerLeft().y)};
+    Point upperRight{std::max(b1.upperRight().x, b2.upperRight().x), std::max(b1.upperRight().y, b2.upperRight().y)};
+
+    return BoundingBox(lowerLeft, upperRight);
+}
+
+inline meshkernel::Point meshkernel::BoundingBox::Centre() const
```
[the remark this answers is not in the document]
Comment (BillSenior) [reply libs/MeshKernel/include/MeshKernel/BoundingBox.hpp:165]: I think we both added the same function, I did not notice after the merge

File: libs/MeshKernel/include/MeshKernel/CurvilinearGrid/CurvilinearGrid.hpp [hunk 3]
```diff
@@ -97,6 +99,14 @@ namespace meshkernel
         /// @param[in] point       The input grid points
         [[nodiscard]] CurvilinearGridNodeIndices GetNodeIndices(Point point);
 
+        Point& GetNode(const UInt i, const UInt j);
```
[the remark this answers is not in the document]
Comment (BillSenior) [reply libs/MeshKernel/include/MeshKernel/CurvilinearGrid/CurvilinearGrid.hpp:102]: added documentation

File: libs/MeshKernel/include/MeshKernel/CurvilinearGrid/CurvilinearGridSnapping.hpp [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+        /// @brief Compute the mesh smoothing factor.
+        ///
+        /// @param [in] snappedNodeIndex Index of the snapped grid point
+        /// @param [in] gridLinePointIndex Current point on the smoothing grid line.
+        virtual double compute(const CurvilinearGridNodeIndices& snappedNodeIndex,
+                               const CurvilinearGridNodeIndices& gridLinePointIndex) const = 0;
+    };
+
+    /// @brief Computes the directional smoothing factor for a point in the grid.
+    class DirectionalSmoothingCalculator : public MeshSmoothingCalculator
```
[the remark this answers is not in the document]
Comment (BillSenior) [reply libs/MeshKernel/include/MeshKernel/CurvilinearGrid/CurvilinearGridSnapping.hpp:59]: Possibly, but I do not know yet. When we are refactoring the curvilinear algorithms we should keep this in mind

File: libs/MeshKernel/src/CurvilinearGrid/CurvilinearGridSnapping.cpp (hunk 1 above)
Comment (BillSenior) [reply libs/MeshKernel/src/CurvilinearGrid/CurvilinearGridSnapping.cpp]: I have moved some of the magic numbers to constants defined in the header.
Its quite difficult to pick a good name for the 10000 and nump

**Review by lucacarniato — APPROVED** [review]

**Pull request merged**